kits/daisy/src/robot/step.cpp

```diff
@@ -0,0 +1,168 @@
+      ik_output,
+      robot_model::EndEffectorPositionObjective(lift_up_));
+    if (ik_output.size() == 0)
+      assert(false);
+
+    // J(1:3, :) \ lift_off_vel;
+    kin.getJEndEffector(ik_output, jacobian_ee);
+    MatrixXd jacobian_part = jacobian_ee.topLeftCorner(3,jacobian_ee.cols());
+    leg_waypoint_vels.block(0, next_pt, num_joints, 1) =
+      jacobian_part.colPivHouseholderQr().solve(lift_off_vel_).eval().cast<double>();
+
+    leg_waypoints.col(next_pt) = ik_output;
+    leg_waypoint_accels.col(next_pt).setZero();
+    
+    leg_times[next_pt] = time_[0];
+    next_pt++;
+  }
+  else // If not initial time, add our current point:
+  {
+    Eigen::VectorXd p(num_joints), v(num_joints), a(num_joints);
+    trajectory_->getState(elapsed, &p, &v, &a);
+
+    // Note -- we will definately have past trajectory here, because it is
+    // created above when (t == start_time_).
+    // TODO: maybe use current position and velocity instead of trajectory ones?
+    leg_waypoints.col(next_pt) = p;
+    leg_waypoint_vels.col(next_pt) = v;
+    leg_waypoint_accels.col(next_pt) = a;
+
+    leg_times[next_pt] = elapsed;
+    next_pt++;
+  }
+
+  // Now, add remaining waypoints
+  if ((time_[1] - elapsed) > ignore_waypoint_threshold_)
+  {
+    kin.solveIK(
+      leg->getSeedAngles(),
+      ik_output,
+      robot_model::EndEffectorPositionObjective(mid_step_1_));
+    leg_waypoints.col(next_pt) = ik_output;
+    leg_waypoint_vels.col(next_pt).setConstant(std::numeric_limits<double>::quiet_NaN());
+    leg_waypoint_accels.col(next_pt).setConstant(std::numeric_limits<double>::quiet_NaN());
+    leg_times[next_pt] = time_[1];
+    next_pt++;
+  }
+  if ((time_[2] - elapsed) > ignore_waypoint_threshold_)
+  {
+    kin.solveIK(
+      leg->getSeedAngles(),
+      ik_output,
+      robot_model::EndEffectorPositionObjective(touch_down_));
+    // J(1:3, :) \ stance_vel;
+    kin.getJEndEffector(ik_output, jacobian_ee);
+    MatrixXd jacobian_part = jacobian_ee.topLeftCorner(3,jacobian_ee.cols());
+    leg_waypoint_vels.block(0, next_pt, num_joints, 1) =
+      jacobian_part.colPivHouseholderQr().solve(leg->getStanceVelXYZ()).eval().cast<double>();
+
+    leg_waypoints.col(next_pt) = ik_output;
+    leg_waypoint_accels.col(next_pt).setZero();
+    leg_times[next_pt] = time_[2];
+    next_pt++;
+  }
+/*
+  if ((time_[3] - elapsed) > ignore_waypoint_threshold_)
+  {
+    kin.solveIK(touch_down_, leg->getSeedAngles(), ik_output);
+    // TODO: get Jacobian, use this to get final vel.
+    leg_waypoints.col(next_pt) = ik_output;
+    leg_waypoint_vels.col(next_pt).setZero();
+    leg_waypoint_accels.col(next_pt).setConstant(std::numeric_limits<double>::quiet_NaN());
+    leg_times.push_back(time_[3]);
+    next_pt++;
+  }
+*/
+  int num_pts = next_pt;
+  // TODO: don't resize, but just be smarter about passing in partial vectors
+  // or creating the right size in the first place!
+  leg_waypoint_vels.conservativeResize(num_joints, num_pts);
```

Comment:
Or better yet, revamp our API to allow for known matrix sizes at compile time (e.g. `Eigen::Matrix<double, 6, 3>`). This is a pain point I saw countless times while porting Igor to C++

Reply:
I'll flag this as an issue on the C++ API